File: backend/routes/receitas_routes.py
from flask import Blueprint, request
from database.db_setup import get_db
from utils.responses import json_response
from utils.receitas_utils import buscar_receitas, normalizar_texto
from routes.auth_routes import token_required

receitas_bp = Blueprint('receitas', __name__)

@receitas_bp.route('/receitas', methods=['GET'])
def listar_receitas():
    """Busca receitas por ingredientes (não requer autenticação)"""
    ingredientes_param = request.args.get('ingredientes', '').strip()
    
    if not ingredientes_param:
        return json_response({'erro': 'Parâmetro "ingredientes" obrigatório'}, 400)
    
    ingredientes = [i.strip() for i in ingredientes_param.split(',') if i.strip()]
    
    if not ingredientes:
        return json_response({'erro': 'Nenhum ingrediente válido'}, 400)
    
    pagina = max(1, request.args.get('pagina', 1, type=int))
    offset = (pagina - 1) * 10
    
    conn = get_db()
    if not conn:
        return json_response({'erro': 'Erro ao conectar no banco'}, 500)
    
    resultados = buscar_receitas(conn, ingredientes)
    conn.close()
    
    total = len(resultados)
    receitas_pagina = resultados[offset:offset+10]
    
    return json_response({
        'sucesso': True,
        'ingredientes': ingredientes,
        'pagina': pagina,
        'total': total,
        'mostrando': len(receitas_pagina),
        'tem_mais': (offset + 10) < total,
        'receitas': receitas_pagina
    })

@receitas_bp.route('/receitas/<int:recipe_id>', methods=['GET'])
def detalhes_receita(recipe_id):
    """Retorna os detalhes de uma receita"""
    conn = get_db()
    if not conn:
        return json_response({'erro': 'Erro ao conectar no banco'}, 500)
    
    try:
        cursor = conn.cursor()
        
        # Busca os dados principais da receita
        cursor.execute("""
            SELECT id, titulo, nota, avaliacoes, autor, 
                   tempo_preparo, link, imagem, descricao, informacoes_adicionais
            FROM recipes
            WHERE id = ?
        """, (recipe_id,))
        
        row = cursor.fetchone()
        
        if not row:
            conn.close()
            return json_response({'erro': 'Receita não encontrada'}, 404)
        
        receita = dict(row)
        
        # Busca os ingredientes
        cursor.execute("""
            SELECT item
            FROM ingredients
            WHERE recipe_id = ?
            ORDER BY id
        """, (recipe_id,))
        
        receita['ingredientes'] = [row['item'] for row in cursor.fetchall()]
        
        # Busca os passos do modo de preparo
        cursor.execute("""
            SELECT text
            FROM recipe_steps
            WHERE recipe_id = ?
            ORDER BY position
        """, (recipe_id,))
        
        receita['modo_preparo'] = [row['text'] for row in cursor.fetchall()]
        
        conn.close()
        
        return json_response({
            'sucesso': True,
            'receita': receita
        })
        
    except Exception as e:
        conn.close()
        print(f"Error: {e}")
        return json_response({'erro': 'Erro ao buscar receita'}, 500)

@receitas_bp.route('/receitas/<int:recipe_id>/favoritar', methods=['POST'])
@token_required
def favoritar_receita(current_user_id, recipe_id):
    """Adiciona ou remove uma receita dos favoritos"""
    conn = get_db()
    if not conn:
        return json_response({'erro': 'Erro ao conectar no banco'}, 500)
    
    try:
        cursor = conn.cursor()
        
        # Verifica se a receita existe
        cursor.execute("SELECT id FROM recipes WHERE id = ?", (recipe_id,))
        if not cursor.fetchone():
            conn.close()
            return json_response({'erro': 'Receita não encontrada'}, 404)
        
        # Verifica se já está favoritada
        cursor.execute("""
            SELECT id FROM favoritos
            WHERE user_id = ? AND recipe_id = ?
        """, (current_user_id, recipe_id))
        
        favorito = cursor.fetchone()
        
        if favorito:
            # Remove dos favoritos
            cursor.execute("""
                DELETE FROM favoritos
                WHERE user_id = ? AND recipe_id = ?
            """, (current_user_id, recipe_id))
            conn.commit()
            conn.close()
            return json_response({
                'sucesso': True,
                'mensagem': 'Receita removida dos favoritos',
                'favoritado': False
            })
        else:
            # Adiciona aos favoritos
            cursor.execute("""
                INSERT INTO favoritos (user_id, recipe_id)
                VALUES (?, ?)
            """, (current_user_id, recipe_id))
            conn.commit()
            conn.close()
            return json_response({
                'sucesso': True,
                'mensagem': 'Receita adicionada aos favoritos',
                'favoritado': True
            })
        
    except Exception as e:
        conn.close()
        print(f"Erro ao favoritar: {e}")
        return json_response({'erro': 'Erro ao favoritar receita'}, 500)

@receitas_bp.route('/receitas/favoritos', methods=['GET'])
@token_required
def listar_favoritos(current_user_id):
    """Lista todas as receitas favoritas do usuário com paginação"""
    conn = get_db()
    if not conn:
        return json_response({'erro': 'Erro ao conectar no banco'}, 500)
    
    try:
        # Paginação
        pagina = max(1, request.args.get('pagina', 1, type=int))
        offset = (pagina - 1) * 10
        
        cursor = conn.cursor()
        
        # Conta total de favoritos
        cursor.execute("""
            SELECT COUNT(*) as total
            FROM favoritos
            WHERE user_id = ?
        """, (current_user_id,))
        
        total = cursor.fetchone()['total']
        
        # Busca favoritos com paginação
        cursor.execute("""
            SELECT r.id, r.titulo, r.nota, r.avaliacoes, r.autor,
                   r.tempo_preparo, r.link, r.imagem, f.data_favoritado
            FROM favoritos f
            JOIN recipes r ON r.id = f.recipe_id
            WHERE f.user_id = ?
            ORDER BY f.data_favoritado DESC
            LIMIT 10 OFFSET ?
        """, (current_user_id, offset))
        
        favoritos = [dict(row) for row in cursor.fetchall()]
        conn.close()
        
        return json_response({
            'sucesso': True,
            'pagina': pagina,
            'total': total,
            'mostrando': len(favoritos),
            'tem_mais': (offset + 10) < total,
            'receitas': favoritos
        })
        
    except Exception as e:
        conn.close()
        print(f"Erro ao buscar favoritos: {e}")
        return json_response({'erro': 'Erro ao buscar favoritos'}, 500)

@receitas_bp.route('/receitas/minhas', methods=['GET'])
@token_required
def listar_minhas_receitas(current_user_id):
    """Lista todas as receitas enviadas pelo usuário com paginação"""
    conn = get_db()
    if not conn:
        return json_response({'erro': 'Erro ao conectar no banco'}, 500)
    
    try:
        # Paginação
        pagina = max(1, request.args.get('pagina', 1, type=int))
        offset = (pagina - 1) * 10
        
        cursor = conn.cursor()
        
        # Conta total de receitas
        cursor.execute("""
            SELECT COUNT(*) as total
            FROM receitas_usuario
            WHERE user_id = ?
        """, (current_user_id,))
        
        total = cursor.fetchone()['total']
        
        # Busca receitas com paginação
        cursor.execute("""
            SELECT id, titulo, descricao, tempo_preparo, imagem, data_criacao
            FROM receitas_usuario
            WHERE user_id = ?
            ORDER BY data_criacao DESC
            LIMIT 10 OFFSET ?
        """, (current_user_id, offset))
        
        receitas = [dict(row) for row in cursor.fetchall()]
        conn.close()
        
        return json_response({
            'sucesso': True,
            'pagina': pagina,
            'total': total,
            'mostrando': len(receitas),
            'tem_mais': (offset + 10) < total,
            'receitas': receitas
        })
        
    except Exception as e:
        conn.close()
        print(f"Erro ao buscar receitas: {e}")
        return json_response({'erro': 'Erro ao buscar receitas'}, 500)

@receitas_bp.route('/receitas', methods=['POST'])
@token_required
def criar_receita(current_user_id):
    """Cria uma nova receita do usuário"""
    data = request.get_json()
    
    if not data:
        return json_response({'erro': 'Dados não fornecidos'}, 400)
    
    titulo = data.get('titulo', '').strip()
    descricao = data.get('descricao', '').strip()
    tempo_preparo = data.get('tempo_preparo', '').strip()
    ingredientes = data.get('ingredientes', [])
    modo_preparo = data.get('modo_preparo', [])
    imagem = data.get('imagem', '').strip()
    
    # Validações
    if not titulo or len(titulo) < 3:
        return json_response({'erro': 'Título deve ter pelo menos 3 caracteres'}, 400)
    
    if not ingredientes or len(ingredientes) < 1:
        return json_response({'erro': 'Adicione pelo menos 1 ingrediente'}, 400)
    
    if not modo_preparo or len(modo_preparo) < 1:
        return json_response({'erro': 'Adicione pelo menos 1 passo no modo de preparo'}, 400)
    
    conn = get_db()
    if not conn:
        return json_response({'erro': 'Erro ao conectar no banco'}, 500)
    
    try:
        cursor = conn.cursor()
        
        # Converte listas para JSON strings
        ingredientes_json = '\n'.join(ingredientes)
        modo_preparo_json = '\n'.join(modo_preparo)
        
        cursor.execute("""
            INSERT INTO receitas_usuario 
            (user_id, titulo, descricao, tempo_preparo, ingredientes, modo_preparo, imagem)
            VALUES (?, ?, ?, ?, ?, ?, ?)
        """, (current_user_id, titulo, descricao, tempo_preparo, 
              ingredientes_json, modo_preparo_json, imagem))
        
        receita_id = cursor.lastrowid
        conn.commit()
        conn.close()
        
        return json_response({
            'sucesso': True,
            'mensagem': 'Receita criada com sucesso',
            'receita_id': receita_id
        }, 201)
        
    except Exception as e:
        conn.close()
        print(f"Erro ao criar receita: {e}")
        return json_response({'erro': 'Erro ao criar receita'}, 500)

@receitas_bp.route('/receitas/minhas/<int:receita_id>', methods=['GET'])
@token_required
def detalhes_minha_receita(current_user_id, receita_id):
    """Retorna os detalhes de uma receita do usuário"""
    conn = get_db()
    if not conn:
        return json_response({'erro': 'Erro ao conectar no banco'}, 500)
    
    try:
        cursor = conn.cursor()
        cursor.execute("""
            SELECT id, titulo, descricao, tempo_preparo, ingredientes, 
                   modo_preparo, imagem, data_criacao
            FROM receitas_usuario
            WHERE id = ? AND user_id = ?
        """, (receita_id, current_user_id))
        
        row = cursor.fetchone()
        conn.close()
        
        if not row:
            return json_response({'erro': 'Receita não encontrada'}, 404)
        
        receita = dict(row)
        receita['ingredientes'] = receita['ingredientes'].split('\n') if receita['ingredientes'] else []
        receita['modo_preparo'] = receita['modo_preparo'].split('\n') if receita['modo_preparo'] else []
        
        return json_response({
            'sucesso': True,
            'receita': receita
        })
        
    except Exception as e:
        conn.close()
        print(f"Erro ao buscar receita: {e}")
        return json_response({'erro': 'Erro ao buscar receita'}, 500)

@receitas_bp.route('/receitas/surpresa', methods=['GET'])
def receita_surpresa():
    """Sorteia uma receita aleatória do banco de dados"""
    conn = get_db()
    if not conn:
        return json_response({'erro': 'Erro ao conectar no banco'}, 500)
    
    try:
        cursor = conn.cursor()
        
        # Conta total de receitas
        cursor.execute("SELECT COUNT(*) as total FROM recipes")
        total = cursor.fetchone()['total']
        
        if total == 0:
            conn.close()
            return json_response({'erro': 'Nenhuma receita disponível'}, 404)
        
        # Sorteia uma receita aleatória
        cursor.execute("""
            SELECT id, titulo, nota, avaliacoes, autor, 
                   tempo_preparo, link, imagem, descricao, informacoes_adicionais
            FROM recipes
            ORDER BY RANDOM()
            LIMIT 1
        """)
        
        row = cursor.fetchone()
        
        if not row:
            conn.close()
            return json_response({'erro': 'Erro ao sortear receita'}, 500)
        
        receita = dict(row)
        recipe_id = receita['id']
        
        # Busca os ingredientes
        cursor.execute("""
            SELECT item
            FROM ingredients
            WHERE recipe_id = ?
            ORDER BY id
        """, (recipe_id,))
        
        receita['ingredientes'] = [row['item'] for row in cursor.fetchall()]
        
        # Busca os passos do modo de preparo
        cursor.execute("""
            SELECT text
            FROM recipe_steps
            WHERE recipe_id = ?
            ORDER BY position
        """, (recipe_id,))
        
        receita['modo_preparo'] = [row['text'] for row in cursor.fetchall()]
        
        conn.close()
        
        return json_response({
            'sucesso': True,
            'mensagem': 'Receita surpresa! 🎉',
            'receita': receita
        })
        
    except Exception as e:
        conn.close()
        print(f"Erro ao sortear receita: {e}")
        return json_response({'erro': 'Erro ao sortear receita'}, 500)

File: backendNode/src/routes/receitasRoutes.js
const express = require('express');
const { getUsersDb, getRecipesDb, runQuery, getQuery, allQuery } = require('../database/dbSetup');
const { jsonResponse } = require('../utils/responses');
const { buscarReceitas } = require('../utils/receitasUtils');
const { tokenRequired } = require('../middleware/authMiddleware');

const router = express.Router();

/**
 * GET /receitas
 * Busca receitas por ingredientes (não requer autenticação)
 */
router.get('/receitas', async (req, res) => {
  try {
    const ingredientesParam = (req.query.ingredientes || '').trim();

    if (!ingredientesParam) {
      return jsonResponse(res, { erro: 'Parâmetro "ingredientes" obrigatório' }, 400);
    }

    const ingredientes = ingredientesParam
      .split(',')
      .map(i => i.trim())
      .filter(i => i.length > 0);

    if (ingredientes.length === 0) {
      return jsonResponse(res, { erro: 'Nenhum ingrediente válido' }, 400);
    }

    const pagina = Math.max(1, parseInt(req.query.pagina) || 1);
    const offset = (pagina - 1) * 10;

    const db = await getRecipesDb();
    const resultados = await buscarReceitas(db, ingredientes);
    db.close();

    const total = resultados.length;
    const receitasPagina = resultados.slice(offset, offset + 10);

    return jsonResponse(res, {
      sucesso: true,
      ingredientes: ingredientes,
      pagina: pagina,
      total: total,
      mostrando: receitasPagina.length,
      tem_mais: (offset + 10) < total,
      receitas: receitasPagina
    });
  } catch (error) {
    console.error('Erro ao buscar receitas:', error);
    return jsonResponse(res, { erro: 'Erro ao buscar receitas' }, 500);
  }
});

/**
 * GET /receitas/favoritos
 * Lista todas as receitas favoritas do usuário com paginação
 */
router.get('/receitas/favoritos', tokenRequired, async (req, res) => {
  try {
    const userId = req.userId;
    const pagina = Math.max(1, parseInt(req.query.pagina) || 1);
    const offset = (pagina - 1) * 10;

    const usersDb = await getUsersDb();

    // Conta total de favoritos
    const totalResult = await getQuery(
      usersDb,
      'SELECT COUNT(*) as total FROM favoritos WHERE user_id = ?',
      [userId]
    );

    const total = totalResult.total || 0;

    // Busca favoritos com paginação (apenas ids)
    const favoritosRows = await allQuery(
      usersDb,
      `SELECT recipe_id, data_favoritado
       FROM favoritos
       WHERE user_id = ?
       ORDER BY data_favoritado DESC
       LIMIT 10 OFFSET ?`,
      [userId, offset]
    );

    usersDb.close();

    let receitas = [];

    if (favoritosRows.length > 0) {
      const ids = favoritosRows.map(f => f.recipe_id);
      const placeholders = ids.map(() => '?').join(',');
      const recipesDb = await getRecipesDb();
      const rows = await allQuery(
        recipesDb,
        `SELECT id, titulo, nota, avaliacoes, autor, tempo_preparo, link, imagem
         FROM recipes
         WHERE id IN (${placeholders})`,
        ids
      );
      recipesDb.close();

      const byId = new Map(rows.map(r => [r.id, r]));
      receitas = favoritosRows
        .map(f => {
          const r = byId.get(f.recipe_id);
          if (!r) return null;
          return {
            id: r.id,
            titulo: r.titulo,
            nota: r.nota,
            avaliacoes: r.avaliacoes,
            autor: r.autor,
            tempo_preparo: r.tempo_preparo,
            link: r.link,
            imagem: r.imagem,
            data_favoritado: f.data_favoritado
          };
        })
        .filter(Boolean);
    }

    return jsonResponse(res, {
      sucesso: true,
      pagina: pagina,
      total: total,
      mostrando: receitas.length,
      tem_mais: (offset + 10) < total,
      receitas: receitas
    });
  } catch (error) {
    console.error('Erro ao buscar favoritos:', error);
    return jsonResponse(res, { erro: 'Erro ao buscar favoritos' }, 500);
  }
});

/**
 * GET /receitas/surpresa
 * Sorteia uma receita aleatória do banco de dados
 */
router.get('/receitas/surpresa', async (req, res) => {
  try {
    const db = await getRecipesDb();

    // Conta total de receitas
    const totalResult = await getQuery(
      db,
      'SELECT COUNT(*) as total FROM recipes'
    );

    const total = totalResult.total || 0;

    if (total === 0) {
      db.close();
      return jsonResponse(res, { erro: 'Nenhuma receita disponível' }, 404);
    }

    // Sorteia uma receita aleatória
    const receita = await getQuery(
      db,
      `SELECT id, titulo, nota, avaliacoes, autor, 
              tempo_preparo, link, imagem, descricao, informacoes_adicionais
       FROM recipes
       ORDER BY RANDOM()
       LIMIT 1`
    );

    if (!receita) {
      db.close();
      return jsonResponse(res, { erro: 'Erro ao sortear receita' }, 500);
    }

    const recipeId = receita.id;

    // Busca os ingredientes
    const ingredientes = await allQuery(
      db,
      'SELECT item FROM ingredients WHERE recipe_id = ? ORDER BY id',
      [recipeId]
    );

    receita.ingredientes = ingredientes.map(row => row.item);

    // Busca os passos do modo de preparo
    const modoPreparo = await allQuery(
      db,
      'SELECT text FROM recipe_steps WHERE recipe_id = ? ORDER BY position',
      [recipeId]
    );

    receita.modo_preparo = modoPreparo.map(row => row.text);

    db.close();

    return jsonResponse(res, {
      sucesso: true,
      mensagem: 'Receita surpresa! 🎉',
      receita: receita
    });
  } catch (error) {
    console.error('Erro ao sortear receita:', error);
    return jsonResponse(res, { erro: 'Erro ao sortear receita' }, 500);
  }
});

/**
 * GET /receitas/:id
 * Retorna os detalhes de uma receita
 */
router.get('/receitas/:id', async (req, res) => {
  try {
    const recipeId = parseInt(req.params.id);
    const db = await getRecipesDb();

    // Busca os dados principais da receita
    const receita = await getQuery(
      db,
      `SELECT id, titulo, nota, avaliacoes, autor, 
              tempo_preparo, link, imagem, descricao, informacoes_adicionais
       FROM recipes
       WHERE id = ?`,
      [recipeId]
    );

    if (!receita) {
      db.close();
      return jsonResponse(res, { erro: 'Receita não encontrada' }, 404);
    }

    // Busca os ingredientes
    const ingredientes = await allQuery(
      db,
      'SELECT item FROM ingredients WHERE recipe_id = ? ORDER BY id',
      [recipeId]
    );

    receita.ingredientes = ingredientes.map(row => row.item);

    // Busca os passos do modo de preparo
    const modoPreparo = await allQuery(
      db,
      'SELECT text FROM recipe_steps WHERE recipe_id = ? ORDER BY position',
      [recipeId]
    );

    receita.modo_preparo = modoPreparo.map(row => row.text);

    db.close();

    return jsonResponse(res, {
      sucesso: true,
      receita: receita
    });
  } catch (error) {
    console.error('Erro ao buscar receita:', error);
    return jsonResponse(res, { erro: 'Erro ao buscar receita' }, 500);
  }
});

/**
 * POST /receitas/:id/favoritar
 * Adiciona ou remove uma receita dos favoritos
 */
router.post('/receitas/:id/favoritar', tokenRequired, async (req, res) => {
  try {
    const userId = req.userId;
    const recipeId = parseInt(req.params.id);

    // Verifica se a receita existe no recipes.db
    const recipesDb = await getRecipesDb();
    const receita = await getQuery(
      recipesDb,
      'SELECT id FROM recipes WHERE id = ?',
      [recipeId]
    );
    recipesDb.close();

    if (!receita) {
      return jsonResponse(res, { erro: 'Receita não encontrada' }, 404);
    }

    // Opera no users.db
    const usersDb = await getUsersDb();

    // Verifica se já está favoritada
    const favorito = await getQuery(
      usersDb,
      'SELECT id FROM favoritos WHERE user_id = ? AND recipe_id = ?',
      [userId, recipeId]
    );

    if (favorito) {
      // Remove dos favoritos
      await runQuery(
        usersDb,
        'DELETE FROM favoritos WHERE user_id = ? AND recipe_id = ?',
        [userId, recipeId]
      );

      usersDb.close();
      return jsonResponse(res, {
        sucesso: true,
        mensagem: 'Receita removida dos favoritos',
        favoritado: false
      });
    } else {
      // Adiciona aos favoritos
      await runQuery(
        usersDb,
        'INSERT INTO favoritos (user_id, recipe_id) VALUES (?, ?)',
        [userId, recipeId]
      );

      usersDb.close();
      return jsonResponse(res, {
        sucesso: true,
        mensagem: 'Receita adicionada aos favoritos',
        favoritado: true
      });
    }
  } catch (error) {
    console.error('Erro ao favoritar:', error);
    return jsonResponse(res, { erro: 'Erro ao favoritar receita' }, 500);
  }
});

/**
 * GET /receitas/favoritos
 * Lista todas as receitas favoritas do usuário com paginação
 */
router.get('/receitas/favoritos', tokenRequired, async (req, res) => {
  try {
    const userId = req.userId;
    const pagina = Math.max(1, parseInt(req.query.pagina) || 1);
    const offset = (pagina - 1) * 10;

    const usersDb = await getUsersDb();

    // Conta total de favoritos
    const totalResult = await getQuery(
      usersDb,
      'SELECT COUNT(*) as total FROM favoritos WHERE user_id = ?',
      [userId]
    );

    const total = totalResult.total || 0;

    // Busca favoritos com paginação (apenas ids)
    const favoritosRows = await allQuery(
      usersDb,
      `SELECT recipe_id, data_favoritado
       FROM favoritos
       WHERE user_id = ?
       ORDER BY data_favoritado DESC
       LIMIT 10 OFFSET ?`,
      [userId, offset]
    );

    usersDb.close();

    let receitas = [];

    if (favoritosRows.length > 0) {
      const ids = favoritosRows.map(f => f.recipe_id);
      const placeholders = ids.map(() => '?').join(',');
      const recipesDb = await getRecipesDb();
      const rows = await allQuery(
        recipesDb,
        `SELECT id, titulo, nota, avaliacoes, autor, tempo_preparo, link, imagem
         FROM recipes
         WHERE id IN (${placeholders})`,
        ids
      );
      recipesDb.close();

      const byId = new Map(rows.map(r => [r.id, r]));
      receitas = favoritosRows
        .map(f => {
          const r = byId.get(f.recipe_id);
          if (!r) return null;
          return {
            id: r.id,
            titulo: r.titulo,
            nota: r.nota,
            avaliacoes: r.avaliacoes,
            autor: r.autor,
            tempo_preparo: r.tempo_preparo,
            link: r.link,
            imagem: r.imagem,
            data_favoritado: f.data_favoritado
          };
        })
        .filter(Boolean);
    }

    return jsonResponse(res, {
      sucesso: true,
      pagina: pagina,
      total: total,
      mostrando: receitas.length,
      tem_mais: (offset + 10) < total,
      receitas: receitas
    });
  } catch (error) {
    console.error('Erro ao buscar favoritos:', error);
    return jsonResponse(res, { erro: 'Erro ao buscar favoritos' }, 500);
  }
});

/**
 * GET /receitas/minhas
 * Lista todas as receitas enviadas pelo usuário com paginação
 */
router.get('/receitas/minhas', tokenRequired, async (req, res) => {
  try {
    const userId = req.userId;
    const pagina = Math.max(1, parseInt(req.query.pagina) || 1);
    const offset = (pagina - 1) * 10;

    const db = await getUsersDb();

    // Conta total de receitas
    const totalResult = await getQuery(
      db,
      'SELECT COUNT(*) as total FROM receitas_usuario WHERE user_id = ?',
      [userId]
    );

    const total = totalResult.total || 0;

    // Busca receitas com paginação
    const receitas = await allQuery(
      db,
      `SELECT id, titulo, descricao, tempo_preparo, imagem, data_criacao
       FROM receitas_usuario
       WHERE user_id = ?
       ORDER BY data_criacao DESC
       LIMIT 10 OFFSET ?`,
      [userId, offset]
    );

    db.close();

    return jsonResponse(res, {
      sucesso: true,
      pagina: pagina,
      total: total,
      mostrando: receitas.length,
      tem_mais: (offset + 10) < total,
      receitas: receitas
    });
  } catch (error) {
    console.error('Erro ao buscar receitas:', error);
    return jsonResponse(res, { erro: 'Erro ao buscar receitas' }, 500);
  }
});

/**
 * POST /receitas
 * Cria uma nova receita do usuário
 */
router.post('/receitas', tokenRequired, async (req, res) => {
  try {
    const userId = req.userId;
    const data = req.body;

    if (!data) {
      return jsonResponse(res, { erro: 'Dados não fornecidos' }, 400);
    }

    const titulo = (data.titulo || '').trim();
    const descricao = (data.descricao || '').trim();
    const tempoPreparo = (data.tempo_preparo || '').trim();
    const ingredientes = data.ingredientes || [];
    const modoPreparo = data.modo_preparo || [];
    const imagem = (data.imagem || '').trim();

    // Validações
    if (!titulo || titulo.length < 3) {
      return jsonResponse(res, { erro: 'Título deve ter pelo menos 3 caracteres' }, 400);
    }

    if (!ingredientes || ingredientes.length < 1) {
      return jsonResponse(res, { erro: 'Adicione pelo menos 1 ingrediente' }, 400);
    }

    if (!modoPreparo || modoPreparo.length < 1) {
      return jsonResponse(res, { erro: 'Adicione pelo menos 1 passo no modo de preparo' }, 400);
    }

    const db = await getUsersDb();

    // Converte arrays para strings
    const ingredientesStr = ingredientes.join('\n');
    const modoPreparoStr = modoPreparo.join('\n');

    const result = await runQuery(
      db,
      `INSERT INTO receitas_usuario 
       (user_id, titulo, descricao, tempo_preparo, ingredientes, modo_preparo, imagem)
       VALUES (?, ?, ?, ?, ?, ?, ?)`,
      [userId, titulo, descricao, tempoPreparo, ingredientesStr, modoPreparoStr, imagem]
    );

    db.close();

    return jsonResponse(
      res,
      {
        sucesso: true,
        mensagem: 'Receita criada com sucesso',
        receita_id: result.id
      },
      201
    );
  } catch (error) {
    console.error('Erro ao criar receita:', error);
    return jsonResponse(res, { erro: 'Erro ao criar receita' }, 500);
  }
});

/**
 * GET /receitas/minhas/:id
 * Retorna os detalhes de uma receita do usuário
 */
router.get('/receitas/minhas/:id', tokenRequired, async (req, res) => {
  try {
    const userId = req.userId;
    const receitaId = parseInt(req.params.id);
    const db = await getUsersDb();

    const receita = await getQuery(
      db,
      `SELECT id, titulo, descricao, tempo_preparo, ingredientes, 
              modo_preparo, imagem, data_criacao
       FROM receitas_usuario
       WHERE id = ? AND user_id = ?`,
      [receitaId, userId]
    );

    db.close();

    if (!receita) {
      return jsonResponse(res, { erro: 'Receita não encontrada' }, 404);
    }

    receita.ingredientes = receita.ingredientes ? receita.ingredientes.split('\n') : [];
    receita.modo_preparo = receita.modo_preparo ? receita.modo_preparo.split('\n') : [];

    return jsonResponse(res, {
      sucesso: true,
      receita: receita
    });
  } catch (error) {
    console.error('Erro ao buscar receita:', error);
    return jsonResponse(res, { erro: 'Erro ao buscar receita' }, 500);
  }
});

module.exports = router;
